File: app/web/src/lib/wsClient.ts
// Real millfolio client over WebSocket — the production transport (see
// ../../../server/STREAMING.md). One WS connection per session: send `ask`,
// receive a stream of ServerEvents, answer any `approval-request` with
// `approve`/`reject` on the same socket. Same MillfolioClient interface as the
// mock, so the UI is identical either way.

import type {
  ClientMessage,
  ServerEvent,
  Session,
  MillfolioClient,
} from "./protocol";

// crypto.randomUUID() throws in a non-secure context / older Safari; fall back so
// opening a session never throws.
function safeId(): string {
  try {
    if (typeof crypto !== "undefined" && crypto.randomUUID) return crypto.randomUUID();
  } catch {}
  return `ask-${Date.now()}-${Math.floor(Math.random() * 1e9)}`;
}

class WsSession implements Session {
  private ws: WebSocket;
  private ready = false;
  private pending: ClientMessage[] = [];
  private done = false; // a terminal event (final message or error) was delivered

  constructor(
    url: string,
    text: string,
    private onEvent: (e: ServerEvent) => void,
    demoToken?: string,
  ) {
    this.ws = new WebSocket(url);

    this.ws.onopen = () => {
      this.ready = true;
      // demo_token: the demo bot gate echoes the Turnstile-minted token on the ask
      // frame (server on_connect validates it). Omitted (harmless) outside the demo.
      this.put({ type: "ask", id: safeId(), text, ...(demoToken ? { demo_token: demoToken } : {}) });
      for (const m of this.pending) this.write(m);
      this.pending = [];
    };

    this.ws.onmessage = (ev) => {
      try {
        const e = JSON.parse(ev.data) as ServerEvent;
        // The server closes the socket right after the final message/error; mark
        // the session done so the close below isn't reported as a drop.
        if (e.type === "message" || e.type === "error") this.done = true;
        this.onEvent(e);
      } catch {
        this.onEvent({ type: "error", message: "malformed event from server" });
      }
    };

    this.ws.onerror = () => {
      if (this.done) return;
      this.done = true;
      this.onEvent({ type: "error", message: `Can't reach the server at ${url}.` });
    };

    // If the socket closes BEFORE a final answer (e.g. the server was restarted
    // mid-request, as during a release), surface it instead of hanging silently.
    this.ws.onclose = () => {
      if (this.done) return;
      this.done = true;
      this.onEvent({
        type: "error",
        message:
          "The server stopped responding (connection closed) — it may be restarting. Try again in a moment.",
      });
    };
  }

  approve(stepId: string) {
    this.put({ type: "approve", stepId });
  }
  reject(stepId: string, reason?: string) {
    this.put({ type: "reject", stepId, reason });
  }

  private write(m: ClientMessage) {
    this.ws.send(JSON.stringify(m));
  }
  // Buffer until the socket is open, then flush in order.
  private put(m: ClientMessage) {
    if (this.ready && this.ws.readyState === WebSocket.OPEN) this.write(m);
    else this.pending.push(m);
  }
}

export function createWsClient(
  url: string,
  getDemoToken?: () => string,
): MillfolioClient {
  return {
    ask(text, onEvent) {
      return new WsSession(url, text, onEvent, getDemoToken?.());
    },
  };
}

File: app/ios/Millfolio/WsClient.swift
// Real Millfolio client over WebSocket — the production transport (see
// ../../server/STREAMING.md). One WS connection per session: send `ask`, receive
// a stream of ServerEvents, answer any `approval-request` with `approve`/`reject`
// on the same socket. Same MillfolioClient interface as the mock, so the UI is
// identical either way. Mirrors web/src/lib/wsClient.ts.

import Foundation

@MainActor
final class WsSession: Session {
    private let task: URLSessionWebSocketTask
    private let url: URL
    private let onEvent: (ServerEvent) -> Void
    private let encoder = JSONEncoder()
    private var closed = false

    init(url: URL, text: String, onEvent: @escaping (ServerEvent) -> Void) {
        self.url = url
        self.onEvent = onEvent
        task = URLSession.shared.webSocketTask(with: url)
        task.resume()
        receive()
        // URLSessionWebSocketTask queues sends until the socket is open, so we can
        // emit `ask` immediately (the web client buffers manually for the same effect).
        write(.ask(id: UUID().uuidString, text: text))
    }

    func approve(stepId: String) { write(.approve(stepId: stepId)) }
    func reject(stepId: String, reason: String?) { write(.reject(stepId: stepId, reason: reason)) }

    private func write(_ m: ClientMessage) {
        guard
            let data = try? encoder.encode(m),
            let json = String(data: data, encoding: .utf8)
        else { return }
        task.send(.string(json)) { [weak self] error in
            guard error != nil else { return }
            Task { @MainActor [weak self] in
                self?.fail("cannot reach server at \(self?.url.absoluteString ?? "")")
            }
        }
    }

    private func receive() {
        task.receive { [weak self] result in
            Task { @MainActor [weak self] in
                guard let self, !self.closed else { return }
                switch result {
                case .failure:
                    self.fail("connection to \(self.url.absoluteString) closed")
                case .success(let message):
                    self.handle(message)
                    self.receive()  // keep reading until the server closes the socket
                }
            }
        }
    }

    private func handle(_ message: URLSessionWebSocketTask.Message) {
        let text: String?
        switch message {
        case .string(let s): text = s
        case .data(let d): text = String(data: d, encoding: .utf8)
        @unknown default: text = nil
        }
        guard let data = text?.data(using: .utf8) else { return }
        if let event = try? JSONDecoder().decode(ServerEvent.self, from: data) {
            onEvent(event)
        } else {
            onEvent(.error(message: "malformed event from server"))
        }
    }

    private func fail(_ message: String) {
        guard !closed else { return }
        closed = true
        onEvent(.error(message: message))
    }

    deinit {
        task.cancel(with: .goingAway, reason: nil)
    }
}

@MainActor
struct WsClient: MillfolioClient {
    let url: URL

    func ask(text: String, onEvent: @escaping (ServerEvent) -> Void) -> Session {
        WsSession(url: url, text: text, onEvent: onEvent)
    }
}
